aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,953 @@
+#
+#                           PROCESSING
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# Conveniently add error context to exceptions raised.  Lets us
+# easily say that an error occurred while processing a specific
+# context.
+@contextlib.contextmanager
+def context(msg: str) -> Iterator[None]:
+    try:
+        yield
+    except Exception as e:
+        # TODO: this does the wrong thing with KeyErorr
+        msg = textwrap.indent(msg, '  ')
+        msg = f'{e.args[0]}\n{msg}' if e.args else msg
+        e.args = (msg,) + e.args[1:]
+        raise
+
+# Represent a source location; used for better error reporting
+@dataclass
+class Location:
+    file: str
+    line: int
+
+    def __str__(self) -> str:
+        return "{}:{}".format(self.file, self.line)
+
+# A custom loader for YAML to let us also keep track of line numbers
+# of each entry in the YAML file
+class LineLoader(Loader):
+    def construct_mapping(self, node, deep=False):  # type: ignore
+        mapping = super().construct_mapping(node, deep=deep)  # type: ignore
+        # Add 1 so line numbering starts at 1
+        mapping['__line__'] = node.start_mark.line + 1
+        return mapping
+
+# Parse native_functions.yaml into a sequence of NativeFunctions
+def parse_native_yaml(path: str) -> List[NativeFunction]:
+    with open(path, 'r') as f:
+        es = yaml.load(f, Loader=LineLoader)
+    assert isinstance(es, list)
+    rs: List[NativeFunction] = []
+    for e in es:
+        assert isinstance(e.get('__line__'), int), e
+        loc = Location(path, e['__line__'])
+        funcs = e.get('func')
+        with context(f'in {loc}:\n  {funcs}'):
+            rs.append(NativeFunction.from_yaml(e, loc))
+    return rs
+
+native_functions = parse_native_yaml('aten/src/ATen/native/native_functions.yaml')
+# pprint.pprint([dataclasses.asdict(f) for f in native_functions])
+
+# TODO: TensorOptions argument detection
+# TODO: Extra enforcement of inplace functions having mutable self
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                           CODE GENERATION
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# This is not fully constructed yet, but eventually this will actually
+# do the code generation
+
+TEMPLATE_PATH = "aten/src/ATen/templates"
+
+# When we interpret things as C++ types, there are a bunch of
+# different modalities we have to consider
+#
+# - Return versus argument type
+# - Mutable type (inplace, out argument)
+# - Public API versus internal calling convention versus legacy calling
+#   convention
+#
+# I'm not really sure how to structure this logic yet, but here is a
+# sketch.  This function is ONLY correct for CPUType.h at the moment;
+# I bet I am going to need another parameter before I'm done
+def cpp_type(t: Type, *, mutable: bool, argument: bool, legacy_optional: bool) -> str:
```

Comment:
@bhosmer suggests making this an lvalue versus rvalue distinction (instead of argument)

Reply:
@bhosmer On further reflection, I'm not sure lvalue v rvalue distinction really makes sense here. lvalue and rvalue refers to expressions, but the computation here involves types!